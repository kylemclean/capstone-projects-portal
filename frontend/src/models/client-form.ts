export default interface ClientForm {
    name: string
    about?: string
    image?: string | BlobPart | null
    website_link?: string
    testimonial?: string
}
