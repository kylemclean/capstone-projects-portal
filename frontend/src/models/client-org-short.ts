import ClientOrgType from "./client-org-type"

export default interface ClientOrgShort {
    id: string
    name: string
    image: string | undefined
    type: ClientOrgType
}
