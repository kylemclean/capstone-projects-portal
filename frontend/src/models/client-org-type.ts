enum ClientOrgType {
    Startup = "Startup",
    Nonprofit = "Non-profit",
    Academic = "Academic",
    CSL = "Community Service Learning",
    Other = "Other",
}

export default ClientOrgType
