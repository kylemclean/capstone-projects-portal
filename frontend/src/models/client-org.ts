import UserShort from "./user-short"
import ProjectShort from "./project-short"
import ClientOrgShort from "./client-org-short"

type ClientOrg = ClientOrgShort & {
    about: string
    website_link: string
    reps: UserShort[]
    projects: ProjectShort[]
    testimonial: string
}
export default ClientOrg
