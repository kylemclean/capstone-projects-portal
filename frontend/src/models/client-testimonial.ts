import type ClientOrgShort from "./client-org-short"

type ClientTestimonial = ClientOrgShort & { testimonial: string }
export default ClientTestimonial
