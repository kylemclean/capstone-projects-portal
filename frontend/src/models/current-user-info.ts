import type ClientOrg from "./client-org"
import type Project from "./project"
import type UserShort from "./user-short"

type CurrentUserInfo =
    | {
          logged_in: false
      }
    | ({
          logged_in: true
          is_superuser: boolean
          has_password: boolean
      } & UserShort)

export default CurrentUserInfo

/**
 * Checks if the current user can edit the given client organization.
 * The user must be logged in and either a superuser
 * or a representative of the client organization.
 *
 * @param currentUser the currently logged in user's CurrentUserInfo
 * @param client the client to be edited
 * @returns true if the user can edit the client organization, false otherwise
 */
export const canEditClientOrg = (
    currentUser: CurrentUserInfo,
    client: ClientOrg
): boolean =>
    currentUser.logged_in &&
    (currentUser.is_superuser ||
        client.reps.map((rep) => rep.id).includes(currentUser.id))

/**
 * Checks if the current user can edit the given user profile.
 * The user must be logged in and either a superuser
 * or the user being edited.
 *
 * @param currentUser the currently logged in user's CurrentUserInfo
 * @param user the ID of the user to be edited
 * @returns true if the user can edit the user profile, false otherwise
 */
export const canEditUser = (
    currentUser: CurrentUserInfo,
    userId: string
): boolean =>
    currentUser.logged_in &&
    (currentUser.is_superuser || userId === currentUser.id)

/**
 * Checks if the current user can edit the given project.
 * The user must be logged in and either a superuser,
 * student of the project, TA of the project, or
 * a representative of the project's client organization.
 *
 * @param currentUser the currently logged in user's CurrentUserInfo
 * @param project the project to be edited
 * @returns true if the user can edit the project, false otherwise
 */
export const canEditProject = (
    currentUser: CurrentUserInfo,
    project: Project
): boolean =>
    currentUser.logged_in &&
    (currentUser.is_superuser ||
        project.students
            .map((student) => student.id)
            .includes(currentUser.id) ||
        (project.ta !== null && project.ta.id === currentUser.id) ||
        (project.client_rep !== null &&
            project.client_rep.id === currentUser.id))
