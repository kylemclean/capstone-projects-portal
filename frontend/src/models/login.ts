import CurrentUserInfo from "./current-user-info"

export interface LoginWithEmailAndPasswordRequest {
    email: string
    password: string
}

export type OAuth2Provider = "GitHub"

export interface LoginWithOAuth2ProviderRequest {
    provider: OAuth2Provider
    code: string
}

export type LoginResult =
    | {
          success: true
          token: string
          user: CurrentUserInfo
      }
    | {
          success: false
          error: string
      }

export interface ActivateRequest {
    activationKey: string
    newPassword: string
}

export type ActivateResult = LoginResult

export type ResetPasswordRequest =
    | {
          currentPassword?: string
          newPassword: string
      }
    | {
          resetKey: string
          newPassword: string
      }

export type ResetPasswordResult = LoginResult

export interface RequestPasswordResetRequest {
    email: string
}
