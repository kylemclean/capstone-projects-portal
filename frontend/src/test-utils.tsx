import * as React from "react"
import { useLocation, useParams } from "react-router-dom"
import { screen, waitFor } from "@testing-library/react"
import { useCurrentUser } from "./hooks"

/**
 * Component to display the login status of the user
 */
export const LoginStatus = (): JSX.Element => {
    const currentUser = useCurrentUser()
    return (
        <div>
            {currentUser.logged_in
                ? `Logged in as ${currentUser.name}`
                : "Not logged in"}
        </div>
    )
}

/**
 * Mock home page that shows the login status
 */
export const MockHomePage = (): JSX.Element => (
    <div>
        Home page
        <LoginStatus />
    </div>
)

/**
 * Mock profile page that shows user ID
 */
export const MockProfilePage = (): JSX.Element => {
    const { id } = useParams<{ id: string }>()
    return <div>Profile page for {id}</div>
}

/**
 * Component that displays each query parameter in a list of "key=value" pairs
 * Requires the use of the `useLocation` hook
 */
export const QueryParamsDisplay = (): JSX.Element => {
    const { search } = useLocation()
    const params = new URLSearchParams(search)
    return (
        <div>
            <ul>
                {Array.from(params.entries()).map(([key, value]) => (
                    <li key={key}>
                        {key}={value}
                    </li>
                ))}
            </ul>
        </div>
    )
}

/**
 * Waits for a query parameter `name` to be set to `value`, or unset.
 * Requires a `QueryParamsDisplay` component to be rendered on the screen.
 * @param name the name of the query parameter
 * @param value the value of the query parameter, or `null` if it should not be present
 */
export const waitForQueryParam = async (
    name: string,
    value: string | null
): Promise<void> =>
    waitFor(() => {
        const text = `${name}=${value}`
        if (value !== null) screen.getByText(text)
        else expect(screen.queryByText(text)).toBeNull()
    })

/**
 * Mock login page that shows where it will redirect to after login
 */
export const MockLoginPage = (): JSX.Element => {
    const location = useLocation()
    const next = new URLSearchParams(location.search).get("next") ?? "/"
    return <div>Login page, will redirect to {next}</div>
}
