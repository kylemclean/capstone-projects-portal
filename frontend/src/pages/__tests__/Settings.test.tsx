import * as React from "react"
import { render, waitFor, screen, fireEvent } from "@testing-library/react"
import "@testing-library/jest-dom"
import { MemoryRouter, Route } from "react-router-dom"
import { setupServer } from "msw/node"
import { ResponseComposition, rest } from "msw"
import GlobalStateProvider from "../../global-state/provider"
import State from "../../global-state/state"
import SettingsPage from "../Settings"
import CurrentUserInfo from "../../models/current-user-info"
import { LoginResult } from "../../models/login"
import requestConfig from "../../api/config"
import { MockLoginPage } from "../../test-utils"

/**
 * Renders the Settings page in a router with a GlobalStateProvider.
 * @param initialState initial state to set for the GlobalStateProvider. If undefined, the state will be the default.
 * @returns the return value of the render function
 */
const renderSettingsPage = (initialState?: State) =>
    render(
        <GlobalStateProvider initialState={initialState}>
            <MemoryRouter initialEntries={["/settings"]}>
                <Route path="/settings" exact component={SettingsPage} />
                <Route path="/login" exact component={MockLoginPage} />
                <Route path="/reset-password" exact>
                    Reset password page
                </Route>
            </MemoryRouter>
        </GlobalStateProvider>
    )

const userWithPassword: CurrentUserInfo = {
    logged_in: true,
    is_superuser: false,
    has_password: true,
    id: "1fc66da3-5f63-4dd8-aaac-425a18804cdc",
    name: "Guy User",
    image: "https://example.com/image.jpg",
}

const userWithoutPassword: CurrentUserInfo = {
    logged_in: true,
    is_superuser: false,
    has_password: false,
    id: "1fc66da3-5f63-4dd8-aaac-425a18804cdc",
    name: "Guy User",
    image: "https://example.com/image.jpg",
}

const userReturnedAfterLoggingOutOtherSessions: CurrentUserInfo = {
    logged_in: true,
    is_superuser: false,
    has_password: true,
    id: "1fc66da3-5f63-4dd8-aaac-425a18804cdc",
    name: "Guy User 2",
    image: "https://example.com/image.jpg",
}

const server = setupServer(
    rest.post(
        `${requestConfig.baseURL}/logout-all/`,
        (req, res: ResponseComposition<LoginResult>, ctx) =>
            res(
                ctx.json({
                    success: true,
                    token: "HEREISYOURNEWTOKEN",
                    user: userReturnedAfterLoggingOutOtherSessions,
                })
            )
    )
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())

it("matches snapshot while logged in", async () => {
    const { container } = renderSettingsPage({ currentUser: userWithPassword })
    expect(container).toMatchSnapshot()
})

it("redirects to login page if not logged in", async () => {
    renderSettingsPage()
    screen.getByText("Login page, will redirect to /settings")
})

it("displays correct password information for user with password", async () => {
    renderSettingsPage({ currentUser: userWithPassword })

    screen.getByText("Your account has a password.")
    screen.getByRole("link", { name: "Change Password" })
})

it("displays correct password information for user without password", async () => {
    renderSettingsPage({ currentUser: userWithoutPassword })

    screen.getByText("Your account does not have a password.")
    screen.getByRole("link", { name: "Set Password" })
})

it("navigates user with password to reset password page when clicking change password button", async () => {
    renderSettingsPage({ currentUser: userWithPassword })

    fireEvent.click(screen.getByRole("link", { name: "Change Password" }))
    screen.getByText("Reset password page")
})

it("navigates user without password to reset password page when clicking set password button", async () => {
    renderSettingsPage({ currentUser: userWithoutPassword })

    fireEvent.click(screen.getByRole("link", { name: "Set Password" }))
    screen.getByText("Reset password page")
})

it("handles log out of all other sessions correctly", async () => {
    // Set initial token
    localStorage.setItem("token", "OLDTOKEN")

    // Render the settings page with an arbitrary user
    renderSettingsPage({ currentUser: userWithPassword })

    // Get Logout of All Other Sessions button
    const logoutButton = screen.getByRole("button", {
        name: "Log Out Of All Other Sessions",
    })

    // Click the button
    fireEvent.click(logoutButton)

    // Expect the button to be disabled
    expect(logoutButton).toBeDisabled()

    // Wait for the request to finish
    await waitFor(() =>
        screen.getByText("Successfully logged out of other sessions")
    )

    // Expect the button to be reenabled
    expect(logoutButton).toBeEnabled()

    // Expect the token to be updated
    expect(localStorage.getItem("token")).toBe("HEREISYOURNEWTOKEN")
})

afterAll(() => server.close())
