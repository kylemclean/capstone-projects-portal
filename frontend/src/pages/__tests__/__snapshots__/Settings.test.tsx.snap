// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot while logged in 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd css-xdmu94-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-zitybv"
    >
      <h1
        class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter MuiTypography-gutterBottom css-1trk376-MuiTypography-root"
      >
        Settings
      </h1>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-i0bk1x-MuiPaper-root"
      >
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Account
          </h4>
          <div
            class="MuiBox-root css-19idom"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              Password
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              depth="2"
            >
              Your account has a password.
            </p>
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
              depth="2"
              href="/reset-password"
              tabindex="0"
            >
              Change Password
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </a>
          </div>
          <div
            class="MuiBox-root css-19idom"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              Current Sessions
            </h5>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-bh137t-MuiButtonBase-root-MuiButton-root"
              depth="2"
              tabindex="0"
              type="button"
            >
              Log Out Of All Other Sessions
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
